drivers/common/virtio/virtio.c

```diff
@@ -0,0 +1,339 @@
+	virtio_dev_set_status(vpdev, VIRTIO_CONFIG_STATUS_DRIVER);
+
+	hw->guest_features = VIRTIO_OPS(hw)->get_features(hw);
+	PMD_INIT_LOG(DEBUG,"Guest_features is 0x%"PRIx64,hw->guest_features);
+	return vpdev;
+
+error:
+		if (vpdev) {
+			rte_free(vpdev);
+		}
+		return NULL;
+
+}
+
+void
+virtio_dev_free(struct virtio_pci_dev *vpdev)
+{
+	struct virtio_hw *hw = NULL;
+
+	hw = &vpdev->hw;
+	if (VIRTIO_OPS(hw)->dev_close(hw))
+		PMD_INIT_LOG(ERR, "Failed to close virtio device.");
+	rte_free(vpdev);
+}
+
+uint16_t
+virtio_dev_nr_vq_get(struct virtio_pci_dev *vpdev)
+{
+	struct virtio_hw *hw = &vpdev->hw;
+	return VIRTIO_OPS(hw)->get_queue_num(hw);
+}
+
+void
+virtio_dev_free_queues(struct virtio_pci_dev *vpdev,uint16_t nr_vq)
+{
+	struct virtio_hw *hw = &vpdev->hw;
+	struct virtqueue *vq;
+	uint16_t i;
+
+	if (hw->vqs == NULL)
+		return;
+
+	for (i = 0; i < nr_vq; i++) {
+		vq = hw->vqs[i];
+		if (!vq)
+			continue;
+		rte_free(vq);
+		hw->vqs[i] = NULL;
+	}
+
+	rte_free(hw->vqs);
+	hw->vqs = NULL;
+}
+
+int
+virtio_dev_alloc_queues(struct virtio_pci_dev *vpdev,uint16_t nr_vq)
+{
+	struct virtio_hw *hw = &vpdev->hw;
+	struct virtqueue *vq;
+	uint16_t i;
+
+	hw->vqs = rte_zmalloc(NULL, sizeof(struct virtqueue *) * nr_vq, 0);
+	if (!hw->vqs) {
+		PMD_INIT_LOG(ERR, "Failed to alloc virtio device queues.");
+		return -ENOMEM;
+	}
+
+	for (i = 0; i < nr_vq; i++) {
+		vq = rte_zmalloc_socket(NULL, sizeof(struct virtqueue), RTE_CACHE_LINE_SIZE,
+					VTPCI_DEV(hw)->device.numa_node);
+		if (vq == NULL) {
+			PMD_INIT_LOG(ERR, "Failed to alloc virtio queue.");
+			virtio_dev_free_queues(vpdev,nr_vq);
+			return -ENOMEM;
+		}
+		hw->vqs[i] = vq;
+		vq->hw = hw;
+		vq->vq_queue_index = i;
+	}
+
+	return 0;
+}
+
+void
+virtio_dev_get_features(struct virtio_pci_dev *vpdev,uint64_t *features)
+{
+	struct virtio_hw *hw = NULL;
+
+	hw = &vpdev->hw;
+	*features = hw->guest_features;
+}
+
+uint64_t
+virtio_dev_set_features(struct virtio_pci_dev *vpdev,uint64_t features)
+{
+	struct virtio_hw *hw = NULL;
+
+	hw = &vpdev->hw;
+	return virtio_dev_negotiate_features(hw, features);
+}
+
+int
+virtio_dev_get_notify_area(struct virtio_pci_dev *vpdev, int qid, uint64_t *offset, uint64_t *size)
+{
+	struct virtio_hw *hw = NULL;
+	struct vfio_region_info reg = { .argsz = sizeof(reg) };
+	int ret;
+
+	reg.index = vpdev->notify_bar;
+	ret = ioctl(vpdev->vfio_dev_fd, VFIO_DEVICE_GET_REGION_INFO, &reg);
+	if (ret) {
+		PMD_INIT_LOG(ERR, "Can not get device region info: %s",
+				strerror(errno));
+		return -1;
+	}
+
+	hw = &vpdev->hw;
+	//TO_DO: check whether notify_addr is 4k aligned
+	*offset = (uint8_t*)hw->vqs[qid]->notify_addr - (uint8_t*)VTPCI_DEV(hw)->mem_resource[vpdev->notify_bar].addr + reg.offset;
+	//nofiy area should 4k aligned, so should not just one doorbell size
+	*size = 0x1000;
+	return 0;
+
+}
+
+int
+virtio_dev_set_queue(struct virtio_pci_dev *vpdev, int index, struct virtio_dev_vring_info *vring_info)
```

Comment:
virtio_dev_vring_info  should be const pointer as this function is not supposed to touch it.

Trim to two lines.